### ISSUE TITLE
Harvesting resin from trees with a knife does crash players

### ISSUE THREAD

**Issue by WickedSchnitzel** [TaskaRaine/Ancient-Tools#63]
```Game Version: v1.18.1 (Stable)
29.04.2023 16:13:26: Critical error occurred
Loaded Mods: 
System.ObjectDisposedException: Cannot access a disposed object.
Object name: 'The ThreadLocal object has been disposed.'.
   at System.Threading.ThreadLocal`1.GetValueSlow()
   at Vintagestory.API.Common.CollectibleObject.UpdateAndGetTransitionStatesNative(IWorldAccessor world, ItemSlot inslot) in VintagestoryApi\Common\Collectible\Collectible.cs:line 2283
   at Vintagestory.API.Common.CollectibleObject.AppendPerishableInfoText(ItemSlot inSlot, StringBuilder dsc, IWorldAccessor world) in VintagestoryApi\Common\Collectible\Collectible.cs:line 1650
   at Vintagestory.API.Common.CollectibleObject.GetHeldItemInfo_Patch0(CollectibleObject this, ItemSlot inSlot, StringBuilder dsc, IWorldAccessor world, Boolean withDebugInfo)
   at Vintagestory.API.Common.Block.GetHeldItemInfo(ItemSlot inSlot, StringBuilder dsc, IWorldAccessor world, Boolean withDebugInfo) in VintagestoryApi\Common\Collectible\Block\Block.cs:line 2173
   at Vintagestory.API.Common.ItemStack.GetDescription(IWorldAccessor world, ItemSlot inSlot, Boolean debug) in VintagestoryApi\Common\Collectible\ItemStack.cs:line 403
   at Vintagestory.GameContent.GuiHandbookItemStackPage..ctor(ICoreClientAPI capi, ItemStack stack) in VSSurvivalMod\Systems\Handbook\Gui\GuiHandbookItemStackPage.cs:line 42
   at Vintagestory.GameContent.ModSystemHandbook.onCreatePagesAsync() in VSSurvivalMod\Systems\Handbook\Handbook.cs:line 79
   at Vintagestory.GameContent.GuiDialogHandbook.LoadPages_Async() in VSSurvivalMod\Systems\Handbook\Gui\GuiDialogHandbook.cs:line 489
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.QueueUserWorkItemCallback.System.Threading.IThreadPoolWorkItem.ExecuteWorkItem()
   at System.Threading.ThreadPoolWorkQueue.Dispatch()

Event Log entries containing Vintagestory.exe, the latest 3
==================================
{ TimeGenerated = 29.04.2023 7:52:34, Site = , Source = .NET Runtime, Message = Application: Vintagestory.exe
Framework Version: v4.0.30319
Description: The process was terminated due to an unhandled exception.
Exception Info: System.NullReferenceException
   at Vintagestory.Common.BlockTypeNet.ReadBlockTypePacket(Packet_BlockType, Vintagestory.API.Common.IWorldAccessor, Vintagestory.Common.ClassRegistry)
   at Vintagestory.Client.NoObf.ClientSystemStartup.PopulateBlocks(Vintagestory.API.Common.Block[], Int32, Int32)
   at Vintagestory.Client.NoObf.ClientSystemStartup.LoadBlockTypes()
   at Vintagestory.Client.NoObf.ClientSystemStartup.HandleServerAssets_Step1()
   at System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
   at System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
   at System.Threading.QueueUserWorkItemCallback.System.Threading.IThreadPoolWorkItem.ExecuteWorkItem()
   at System.Threading.ThreadPoolWorkQueue.Dispatch()

 }
--------------
{ TimeGenerated = 02.04.2023 17:12:36, Site = , Source = Application Hang, Message = The program Vintagestory.exe version 1.17.11.0 stopped interacting with Windows and was closed. To see if more information about the problem is available, check the problem history in the Security and Maintenance control panel.

Process ID: 38ac

Start Time: 01d96541602656fe

Termination Time: 36

Application Path: D:\Users\Grigoriewich\AppData\Roaming\Vintagestory165\Vintagestory.exe

Report Id: e8d1cdc0-510f-450f-aed4-5766f0a586d0

Faulting package full name: 

Faulting package-relative application ID: 

Hang type: Unknown
 }
--------------
{ TimeGenerated = 02.04.2023 11:48:25, Site = , Source = .NET Runtime, Message = Application: Vintagestory.exe
Framework Version: v4.0.30319
Description: The process was terminated due to an unhandled exception.
Exception Info: System.ObjectDisposedException
   at System.Threading.ThreadLocal`1[[System.__Canon, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]].GetValueSlow()
   at Vintagestory.API.Common.CollectibleObject.UpdateAndGetTransitionStatesNative(Vintagestory.API.Common.IWorldAccessor, Vintagestory.API.Common.ItemSlot)
   at Vintagestory.API.Common.CollectibleObject.AppendPerishableInfoText(Vintagestory.API.Common.ItemSlot, System.Text.StringBuilder, Vintagestory.API.Common.IWorldAccessor)
   at DynamicClass.Vintagestory.API.Common.CollectibleObject.GetHeldItemInfo_Patch0(Vintagestory.API.Common.CollectibleObject, Vintagestory.API.Common.ItemSlot, System.Text.StringBuilder, Vintagestory.API.Common.IWorldAccessor, Boolean)
   at Vintagestory.API.Common.ItemStack.GetDescription(Vintagestory.API.Common.IWorldAccessor, Vintagestory.API.Common.ItemSlot, Boolean)
   at Vintagestory.GameContent.GuiHandbookItemStackPage..ctor(Vintagestory.API.Client.ICoreClientAPI, Vintagestory.API.Common.ItemStack)
   at Vintagestory.GameContent.GuiDialogHandbook.BuildPages_Async(System.Collections.Generic.List`1<Vintagestory.API.Common.ItemStack>)
   at System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
   at System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
   at System.Threading.QueueUserWorkItemCallback.System.Threading.IThreadPoolWorkItem.ExecuteWorkItem()
   at System.Threading.ThreadPoolWorkQueue.Dispatch()

 }```

**Comment by WickedSchnitzel** [comment]
it is believed that your mod is the cause

**Comment by TaskaRaine** [comment]
What version of Ancient Tools are you using? o.o 

Also, from this report, it looks like you're running Vintage Story 1.17.11? If that's the case, make sure you're using Ancient Tools 1.5.1 or earlier. 1.5.1 is the latest version built for Vintage Story 1.17

I've been having reports of crashes with resin in Vintage Story 1.18, but this crash is not at all similar to that. This one seems to be related to something handbook and/or UI related? I've launched my 1.17.11 game with Ancient Tools 1.5.1 to test to see if placing/harvesting the directional resin blocks causes crashes, and also carved and accelerated time to harvest and...well... still no problems on single player at least.

The crashes that others have reported always points to line 100 in the vanilla Harvestable behavior, probably in VS 1.18. Unfortunately, I've not been able to reproduce that crash and my leading theory is that the crash is caused by moss growing on the leaking logs. Reports of this type of crash actually exist before Ancient Tools even made its way to Vintage Story 1.18, so it could be possible that this mod may be exacerbating an existing vanilla issue. 

Something to note: Ancient Tools resin harvesting only uses vanilla behaviors implemented through JSON on blocks that mimic vanilla blocks to ensure the carved face is looking towards the player that carved it. It does not implement unique code for resin harvesting. I did implement a possible solution today for some crashes via custom code, but that's just a hope and a prayer.

If the mod most definitely isn't working and should be please please offer up some replication steps if you have them. This isn't something I can solve without being able to replicate it.

**Comment by WickedSchnitzel** [comment]
Our server is on 1.18.1 - but the client of that player may be on 1.17, ill
have to check that.

Taska Raine ***@***.***> schrieb am So., 30. Apr. 2023, 00:24:

**Comment by RogueRaiden** [comment]
apparently this was a vanilla issue, and should be fixed as of 1.18.3 i believe

**Comment by TaskaRaine** [comment]
Yup yup, should be fixed with the vanilla game
